test/metric_learn_test.py

```diff
@@ -378,211 +380,97 @@ def test_loss_grad_lbfgs(self):
     X, y = make_classification(random_state=rng)
     L = rng.randn(rng.randint(1, X.shape[1] + 1), X.shape[1])
     lmnn = LMNN()
+    lmnn.n_iter_ = 0
+    lmnn.n_neighbors_ = lmnn.n_neighbors
 
-    k = lmnn.k
-    reg = lmnn.regularization
-
-    X, y = lmnn._prepare_inputs(X, y, dtype=float,
-                                ensure_min_samples=2)
+    X, y = lmnn._prepare_inputs(X, y, dtype=float, ensure_min_samples=2)
     num_pts, n_components = X.shape
-    unique_labels, label_inds = np.unique(y, return_inverse=True)
-    lmnn.labels_ = np.arange(len(unique_labels))
+    unique_labels, y_inverse = np.unique(y, return_inverse=True)
+    classes = np.arange(len(unique_labels))
     lmnn.components_ = np.eye(n_components)
 
-    target_neighbors = lmnn._select_targets(X, label_inds)
+    target_neighbors = lmnn._select_target_neighbors(X, y_inverse, classes)
 
     # sum outer products
-    dfG = _sum_outer_products(X, target_neighbors.flatten(),
-                              np.repeat(np.arange(X.shape[0]), k))
+    tn_graph = _make_knn_graph(target_neighbors)
+    pull_loss_grad_m = _sum_weighted_outer_products(X, tn_graph)
 
-    # initialize L
-    def loss_grad(flat_L):
-      return lmnn._loss_grad(X, flat_L.reshape(-1, X.shape[1]), dfG,
-                             k, reg, target_neighbors, label_inds)
+    kwargs = {
+        'classes':  classes,
+        'target_neighbors': target_neighbors,
+        'pull_loss_grad_m': pull_loss_grad_m,
+    }
 
-    def fun(x):
-      return loss_grad(x)[1]
+    def fun(L):
+        return lmnn._loss_grad_lbfgs(L, X, y, **kwargs)[0]
 
-    def grad(x):
-      return loss_grad(x)[0].ravel()
+    def grad(L):
+        return lmnn._loss_grad_lbfgs(L, X, y, **kwargs)[1]
 
     # compute relative error
-    epsilon = np.sqrt(np.finfo(float).eps)
-    rel_diff = (check_grad(fun, grad, L.ravel()) /
-                np.linalg.norm(approx_fprime(L.ravel(), fun, epsilon)))
+    rel_diff = check_grad(fun, grad, L.ravel()) / np.linalg.norm(grad(L))
     np.testing.assert_almost_equal(rel_diff, 0., decimal=5)
 
 
-def test_loss_func(capsys):
```

Comment:
I think I removed it because the checks that can remain are covered by other tests, such as the `test_toy_ex_lmnn()`. Sure, if you think there's still something that's not covered already, I'll add it back.

Reply:
Yes, I think `test_toy_ex_lmnn` is good because we can compute the result by hand, but it's maybe too light and simple, I think it would be good to have a more "real" example, with many random points (even more than 10 like in the original example, maybe like 40)
For that I think it would be quite simple to adapt the previous example, just checking the first iteration of loss and gradient: the whole last part of the original test with callbacks etc (not applicable here because of lbfgs) could be removed, let me know what you think

Reply:
Sure, sounds good.